prov/gni/src/gnix_mr.c

```diff
@@ -295,14 +360,101 @@ DIRECT_FN int gnix_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
 		return -FI_EOPNOTSUPP;
 
 	if (attr->iov_count == 1)
-		return __mr_reg(fid, attr->mr_iov[0].iov_base, 
+		return _gnix_mr_reg(fid, attr->mr_iov[0].iov_base, 
 			attr->mr_iov[0].iov_len, attr->access, attr->offset,
-			attr->requested_key, flags, mr, attr->context);
+			attr->requested_key, flags, mr, attr->context, 
+			USER_REGISTRATION);
 	
 	/* regv limited to one iov at this time */
 	return -FI_EOPNOTSUPP;
 }
 
+static int __gnix_mr_refresh(struct gnix_fid_mem_desc *desc, uint64_t addr, uint64_t len)
+{
+		gni_return_t grc;
+
+
+	fastlock_acquire(&desc->nic->lock);
+	GNIX_INFO(FI_LOG_MR, "hndl=%p addr=%p len=%d cq=%p flags=%x vmdh_index=%d mem_hndl=%p\n",
+		desc->nic->gni_nic_hndl, addr, len, NULL, GNI_MEM_UPDATE_REGION,
+		desc->mr_fid.key, &desc->mem_hndl);
+	grc = GNI_MemRegister(desc->nic->gni_nic_hndl, addr,
+		len, NULL, GNI_MEM_UPDATE_REGION,
+		desc->mr_fid.key, &desc->mem_hndl);
+	assert(grc == FI_SUCCESS); /* TODO: handle error case */
+
+	fastlock_release(&desc->nic->lock);
+
+	return (grc != 0) ? -FI_EINVAL : FI_SUCCESS;
+}
+
+static int __gnix_mr_refresh_iov(struct fid *fid, void *arg)
+{
+	struct fi_mr_modify *modify = (struct fi_mr_modify *) arg;
+	int ret = FI_SUCCESS;
+	struct gnix_fid_mem_desc *desc;
+	uint64_t aligned_addr;
+	uint64_t aligned_len;
+	uint64_t addr; 
+	uint64_t len;
+	int i;
+	
+	GNIX_TRACE(FI_LOG_MR, "\n");
+	
+	desc = container_of(fid, struct gnix_fid_mem_desc, mr_fid);
+
+	/* assume that no one is going to attempt to update a MR at the 
+		same time that they might try to deregister a mr 
+		
+	   For the record, that would be REALLY silly application behavior */
+
+	if (!modify->attr.mr_iov || modify->attr.iov_count == 0)
+		return -FI_EINVAL;
+
+	for (i = 0; i < modify->attr.iov_count; i++) {
+		addr = (uint64_t) modify->attr.mr_iov[i].iov_base;
+		len = (uint64_t) modify->attr.mr_iov[i].iov_len;
+
+		if (!(desc->addr <= (uint64_t) addr && (((uint64_t) addr + len) <= (desc->addr + desc->len))))
+			return -FI_EINVAL;
+
+		aligned_addr = addr & ~0xfff;
+		aligned_len = addr + len - aligned_addr;
+		aligned_len += (((addr + len) & 0xfff) ? (0x1000 - ((addr + len) & 0xfff)) : 0); 
```

Comment:
Could we do `aligned_len = len + (addr - aligned_addr)` instead? Also, is `aligned_addr + aligned_len` supposed to result in a aligned address?